### PR TITLE
POC GraphNG: upstream data alignment

### PR DESCRIPTION
**Pull request by dprokop**
Maybe a crazy idea, but got inspired by my conversation with @leeoniya 

This PR adds an API to PanelPlugin to enable forcing data alignment in the PanelQueryRunner data pipeline. Not surprisingly it simplifies the GraphNG quite significantly, as we don't need to care about whether or not the alignment needs to happen and when the config should be passed down to uPlot (farewell getDerivedStateFromProps).

This is WIP and only for discussion for now.

### CODE REVIEW

File: public/app/features/dashboard/dashgrid/PanelChrome.tsx [hunk 1]
```diff
@@ -90,10 +90,23 @@ export class PanelChrome extends Component<Props, State> {
       this.setState({ isFirstLoad: false });
     }
 
+    let alignmentOptions;
+
+    if (Boolean(panel.plugin?.meta.alignData) && panel.plugin?.resolveXYDimensions) {
+      alignmentOptions = {
+        withAlignment: Boolean(panel.plugin?.meta.alignData),
+        alignmentOptions: (data: DataFrame[]) => panel.plugin!.resolveXYDimensions!(data, panel.options),
+      };
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/PanelChrome.tsx:99]: nitpick: the naming here feels a bit off, should the alignmentOptions function be getAlignmentMatchers ?

**Review by torkelo — COMMENTED** [review]
Looks like this simplifies a lot of the internals. Not been that involved in the alignment issues so interested to know what @ryantxu and @leeoniya think. 

Think this requires an update here as well: https://github.com/grafana/grafana/blob/master/public/app/features/panel/PanelRenderer.tsx#L27 

And maybe in explore as well?

**Comment by dprokop** [comment]
@torkelo thanks a lot for taking a look! We are just exploring this for now to see how/if this helps at all. It lokos promising, yet there's stil an issue with config/series length mismatch in the streaming scenario that I'm currently investigating

**Review by hugohaggmark — COMMENTED** [review]
I'm afraid I don't know enough about the underlying issue to find anything disturbing. I think this looks good so far.

**Comment by ryantxu** [comment]
This contains a few interesting ideas -- but I am kinda nervous that this feels required to simplify the hook/react madness in GraphNG.  

I *do* think we should add additional flags on `panelData` to indicate if the schema has changed vs the data -- then panels can look to that as a fast hint

**Comment by dprokop** [comment]
> This contains a few interesting ideas -- but I am kinda nervous that this feels required to simplify the hook/react madness in GraphNG.

Yeah, GraphNG is a driver here, yet mind that this does not influence the hooks at all, nor does mitigate any internal Plot.tsx issue. 

What it does tho, it assures the data coming to the panel is always in a "ready" state as required by the panel.

But, I think we can actually skip the data alignment in the PQR and only have the structure and data revisions passed down to the panel. This will be sufficient for us to make a decision about the data update without the super cryptic `getDerivedStateFromProps`. Let's see where this will take us, I'll work on this today.

**Comment by dprokop** [comment]
@ryantxu @torkelo - me and @leeoniya also discussed an alternative approach to this one, when the aligned data gets added to PanelData as a callback. The data passed down to the panel wouldn't change, yet the consumer would get a function for lazy retrieval of aligned data (in a situation when the panel is configured to support that via plugin.json). 

I think this is also an interesting idea, yet I'm not a huge fan of PanelPlugin API changes in this PR(tho, they make few things nicer like less data processing in GraphNG and easier data retrieval i.e. in tooltip just by the index, with no mapping to the original data frame required)

**Comment by dprokop** [comment]
Closing in favor of https://github.com/grafana/grafana/pull/33215

**Pull request closed without merging**